Add `RELEASES.md` (#50577)
moby/moby: RELEASES.md

# 🚀 Releases

This document outlines the Moby project’s release process, including cadence and versioning.

## 📌 Versioning

> Releases of moby modules will be versioned using dotted triples, similar to

Reviewer: It's a module, so not similar to, it _must_ be SemVer (its part of the contract of using modules)

[Semantic Version](http://semver.org/). For the purposes of this document, we
will refer to the respective components of this triple as
`<major>.<minor>.<patch>-[suffix[N]]`. The version number may have additional
information, such as alpha, beta and release candidate qualifications along with
a numeric identifier. Such releases will be considered "pre-releases".

> > [!NOTE]

Reviewer: We can probably create a separate section about "relation to the docker/docker" module.

(or at least, most of this information is/becomes obsolete, and describes a different module)

> v17.03-v20.10 releases were following [CalVer](https://calver.org/).
> v2.0.0 introduces the `github.com/moby/moby/v2` go module after the
> previous v28 release of the `github.com/docker/docker` pseudo-module.

## ⏳ Release cadence

The project will release as quickly as possible while ensuring stability and
timely feature delivery. Maintainers curate a feature roadmap and set milestones
for each release.

> Release dates assigned to milestones are commitments, and any changes will be
> announced to the community via an
> [announcement discussion](https://github.com/moby/moby/discussions/categories/announcements).
> To meet deadlines, some features may be deferred to future releases.

Reviewer 2: Considering how unpredictable things can be, I'd suggest making it a little bit softer:

```suggestion
Release dates assigned to milestones are estimates.
To meet deadlines, some features may be deferred to future releases.
```

We can always change it later to be a stronger commitment when we reach the level of predictability that would allow that.

If your issue or feature is missing from the roadmap, please open a GitHub issue
or comment to request its inclusion.

## 🎯 Major and minor releases

Major and minor releases of moby will be made from main. Releases will be marked

> with GPG signed tags and posted at https://github.com/moby/moby/releases. The

Reviewer 3: Tried SSH key signing for the alpha releases and it worked pretty well. Not a blocker, but if you go back under the hood on this one can updated it to just mentioned the tags will be signed.

> tag will be of the format `v<major>.<minor>.<patch>-[suffix[N]]`.

Reviewer 2: For consistency with previous versions (like https://github.com/moby/moby/releases/tag/v28.3.0-rc.2), let's keep the `.` between the suffix and N:

```suggestion
tag will be of the format `v<major>.<minor>.<patch>-[suffix.[N]]`.
```

After a minor release of the main module, a branch will be created, with the format
`release/<major>.<minor>` from the minor tag. All further patch releases will be
done from that branch. For example, after the release of `v2.1.0`, a branch

> `release/2.1` would be created from that tag. All future patch releases will be

Reviewer: This part (versioning) we still need to look closely at.

Given that we provided NO Go API stability, we can decide to stay on "v2 forever".

At least we must avoid reaching v17, because then we'll collide with v17.12.0, which ISTR go modules considered SemVer (but +incompatible)

Author: Or maybe we just can't count... v15, v16, v29, v30 😆 

I'm hoping being clear about what we consider "supported" helps us define this. We respect can both semver and  take compatibility seriously by communicating what exactly the contract is and recommending against importing unsupported packages. I think before we hit a v3 we could probably just stabilize the Go interface with proper use of internal packages, or even call it v3 at that point.

Reviewer 3: It'll take some time for the Go interface to shake out. Users will provide feedback on what they like and dislike and we will need some flexibility to change it. In the meantime, the client and API are stable and satisfy a majority of users.

> Or maybe we just can't count... v15, v16, v29, v30 😆

It's unlikely I'll see v16 + 1 in my lifetime at least skipping is a simple enough solution. Or maybe Go will have swapped to epoch semver by then.

- `github.com/moby/moby/api`: The API module, which contains the API types and documentation.
- `github.com/moby/moby/client`: The client module, which contains the client library for
  interacting with the Moby API.

## Support Horizon

Support horizons will be defined corresponding to a release branch, identified
by `<major>.<minor>`. Release branches will be in one of several states:

- __*Future*__: An upcoming scheduled release.
- __*Alpha*__: The next scheduled release on the main branch under active development.
- __*Beta*__: The next scheduled release on the main branch under testing. Begins 2-4 weeks before a final release.
- __*RC*__: The next scheduled release on the main branch under final testing and stabilization. Begins 1 week before a final release. For new releases where the source branch is main, the main branch will be in a feature freeze during this phase.
- __*Active*__: The release is a stable branch which is currently supported and accepting patches.
- __*Extended*__: The release branch is only accepting security patches.
- __*End of Life*__: The release branch is no longer supported and no new patches will be accepted.

> Releases will be supported until at least the next _minor_ release. Any minor release

Reviewer: I can do some follow ups, but try to avoid future tense in docs. While this is for sure "for future events", I think most parts still work well (and more to the point) in present tense.

| v2.1        | v1.x    | v1.x   | *Future* |              |             |          |

### Pre-modules releases branches

| Release Branch | Status      | End of Life  | Owner(s)         |
|----------------|-------------|--------------|------------------|
| 28.x           | Active      | v2.0 release | @moby-committers |
| 27.x           | End of Life |              |                  |
| 26.1           | End of Life |              |                  |
| 26.0           | End of Life |              |                  |
| 25.0           | Active      | See [mcr25]  | @corhere         |
| 24.0           | End of Life |              |                  |
| 23.0           | Extended    | See [mcr23]  | @corhere         |

[mcr25]: https://docs.mirantis.com/mcr/25.0/compat-matrix/maintenance-lifecycle.html
[mcr23]: https://docs.mirantis.com/mcr/23.0/compat-matrix/maintenance-lifecycle.html

## API Stability

The following table provides an overview of the components covered by
Moby versions and the stability guarantees provided to users and importers.

| Component        | Status        | Stabilized Version | Links |
|------------------|---------------|--------------------|-------|

> | HTTP API         | Stable        | 1.0                | [Docker Engine API](https://docs.docker.com/reference/api/engine/) |

Reviewer: We may need some wording about currently supported API versions outline that we dropped support for v1.23 and before (and may start moving the lower version).

In general (I added some wording in the API docs/README.md), support for legacy versions (especially really old ones) is "best effort"; some functionality may be missing when we downgrade the daemon or client to an older API version.

Author: I think in the above "Support Horizons" it would make sense to have a separate tables about API versions. I think its less confusing to have more tables with very specific information than trying to fit too much into one table. The current one there should probably only have the main module and we can separate the tables for client and API support and which versions they are related to.

| API Go module    | *In Progress* | 1.x                | [`github.com/moby/moby/api`](https://pkg.go.dev/github.com/moby/moby/api) |
| Client Go module | *In Progress* | 1.0                | [`github.com/moby/moby/client`](https://pkg.go.dev/github.com/moby/moby/client) |
| Daemon Config    | Stable        | 1.0                | [Daemon configuration file](https://docs.docker.com/reference/cli/dockerd/#daemon-configuration-file) |
| Main Go Module   | Unstable      | Unsupported        | [`github.com/moby/moby/v2`](https://pkg.go.dev/github.com/moby/moby/v2) |

> NOTE: While the daemon's HTTP API and configuration file are considered stable, the Go code within the main module

> > is not considered stable. The `github.com/moby/moby/api` and `githhub.com/moby/moby/client` modules are instended

Reviewer: typo "instended"
